### DISCUSSION TITLE
Considerations for container base image selection

### DISCUSSION THREAD

**Discussion by andez2000** [devcontainers/community#23] [Q&A]
I was looking at getting a devcontainer together where I could run `minikube` on a `Python` devcontainer as an experiment.  I generated the `devcontainer.json` using VSCode which produced the following.

```json
{
	"name": "Kubernetes - Minikube-in-Docker",
	"image": "mcr.microsoft.com/devcontainers/base:bullseye",

	"features": {
		"ghcr.io/devcontainers/features/docker-in-docker:2": {
			"enableNonRootDocker": "true",
			"moby": "true"
		},
		"ghcr.io/devcontainers/features/kubectl-helm-minikube:1": {},
		"ghcr.io/devcontainers/features/python:1": {
			"version": "3.11"
		}
	}
}
```

Minikube starts up without a problem.

## Debian Python Bullseye

Now if I switch the image to **`python:3.11-bullseye`.

```json
{
	"name": "Kubernetes - Minikube-in-Docker",
	"image": "python:3.11-bullseye",

	"features": {
		"ghcr.io/devcontainers/features/docker-in-docker:2": {
			"enableNonRootDocker": "true",
			"moby": "true"
		},
		"ghcr.io/devcontainers/features/kubectl-helm-minikube:1": {},
		"ghcr.io/devcontainers/features/python:1": {
			"version": "3.11"
		}
	}
}
```

This yields the following error message:

```
😄  minikube v1.28.0 on Debian 11.6 (docker/amd64)
✨  Automatically selected the docker driver. Other choices: none, ssh
🛑  The "docker" driver should not be used with root privileges. If you wish to continue as root, use --force.
💡  If you are running minikube within a VM, consider using --driver=none:
📘    https://minikube.sigs.k8s.io/docs/reference/drivers/none/

❌  Exiting due to DRV_AS_ROOT: The "docker" driver should not be used with root privileges.
```

I guess there are some subtle differences in the choice of image.  In other experiments I've used the `python:3.11-bullseye` image to add some customisation into a dockerfile without any major issues. 

Ideally I want to control my environment as much as possible, but what considerations do you apply when selecting a devcontainer base image?

Ideally I would like to have:
* Python v3.11
* Minikube

With my choice of VSCode extensions (but the features install those).  It would be cool to omit extension installation in the JSON (if we cannot already).

Where I can route calls from my Host OS Browser through to a FastAPI running in Minikube inside the devcontainer.

![image](https://user-images.githubusercontent.com/16876524/211421806-45a9e860-5fc0-4e77-8739-6318062b060e.png)

**Comment by Chuxel** [comment]
Yeah, a number of developer tools do not like running as root.  This is why all of the pre-built devcontainer images include a non-root user. So the error you see is exactly right.  

- The `mcr.microsoft.com/devcontainers/base:bullseye` image has a non-root user that is used by default.  The container itself runs as root to allow things like docker-in-docker to function, but tools connect as non-root. The image is built using the Dev Container CLI to [include devcontainer.json metadata in an image label](https://containers.dev/implementors/reference/#labels). You can see the devcontainer.json and Dockerfile used to create this image here: https://github.com/devcontainers/images/tree/main/src/base-debian 
- `python:3.11-bullseye` does not have a non-root user at all, so one needs to be created.  You can then use the `remoteUser` property in devcontainer.json to use it.  `mcr.microsoft.com/devcontainers/python:3.11-bullseye` actually extends this image and is like `base` but with python in it. You can see how this is created here: https://github.com/devcontainers/images/tree/main/src/python

The "[ghcr.io/devcontainers/features/common-utils](https://github.com/devcontainers/features/tree/main/src/common-utils)" feature is intended to be an easy way to add a non-root user and tools developers tend to prefer using like man pages. You'll note this is referenced in all the devcontainer.json files used to build the mcr.microsoft.com images. You can then reference this non-root user as the `remoteUser` if you want to keep the container running as root, or `containerUser` if you'd prefer the whole container runs as non-root - though some things may not work as expected in this case.
